workspace_host/interface.js

```diff
@@ -235,11 +252,173 @@ watcher.on('unlinkDir', filename => {
     var key = [filename, true];
     update_queue[key] = {action: 'delete'};
 });
-setInterval(_autoUpdateJobManager, interval * 1000);
+setInterval(_autoUpdateJobManager, config.workspaceHostFileWatchIntervalSec * 1000);
+
+/* Periodic hard-push of files to S3 */
+
+/**
+ * Push all of the contents of a container's home directory to S3.
+ * @param {object} workspace Workspace object, this should contain at least the launch_uuid and id.
+ */
+async function pushContainerContentsToS3(workspace) {
+    const workspacePath = path.join(workspacePrefix, `workspace-${workspace.launch_uuid}`);
+    const settings = _getWorkspaceSettingsAsync(workspace.id);
+    try {
+        await workspaceHelper.uploadDirectoryToS3Async(workspacePath, `${config.workspaceS3Bucket}/workspace-${workspace.id}`, settings.workspace_sync_ignore);
+    } catch (err) {
+        /* Ignore any errors that may occur when the directory doesn't exist */
+        logger.error(`Error uploading directory: ${err}`);
+    }
+}
 
+/**
+ * Push the contents of all running workspaces to S3.  Workspace home directories are uploaded
+ * serially instead of in parallel.
+ */
+async function pushAllRunningContainersToS3() {
+    const result = await sqldb.queryAsync(sql.get_running_workspaces, { instance_id: workspace_server_settings.instance_id });
+    await async.eachSeries(result.rows, async (ws) => {
+        if (ws.state == 'running') {
+            await pushContainerContentsToS3(ws);
+        }
+    });
+}
+setInterval(pushAllRunningContainersToS3, config.workspaceHostForceUploadIntervalSec * 1000);
+
+/* Prune stopped and runaway containers */
+
+/**
+ * Remove any recently stopped containers.  These are identified by having a non-null launch_uuid
+ * and launch_port, but are marked as stopped by the main PrairieLearn instance.
+ */
+async function pruneStoppedContainers() {
+    const instance_id = workspace_server_settings.instance_id;
+    const recently_stopped = await sqldb.queryAsync(sql.get_recently_stopped_workspaces, { instance_id });
+    await async.each(recently_stopped.rows, async (ws) => {
+        try {
+            /* Try to kill, but don't care if it's already dead */
+            const container = await _getDockerContainerByLaunchUuid(ws.launch_uuid);
+            await container.kill();
+            await container.remove();
+        } catch (err) {
+            logger.error(err);
+        }
+        await sqldb.queryAsync(sql.set_workspace_launch_uuid, { workspace_id: ws.id, launch_uuid: null });
+        await sqldb.queryAsync(sql.set_workspace_launch_port, { workspace_id: ws.id, launch_port: null });
```

Comment:
We could also set `workspace_host_id = null`. That way we are saying this WS is not associated with us in any way anymore. Then it won't show up the next time we check for stopped containers, and we can change `get_recently_stopped_workspaces` to get all stopped workspaces, not just recently stopped ones.